List-Method Exercise Edits and List Concept Doc (#2786)
exercism/v3: languages/python/concepts/lists/about.md

                        "Lavender",
                        "Daisy",
                        "Jasmine",
                        "Hydrangea",
                        "Hyacinth",
                        "Peony",
                        "Dahlia",
                        "Dandelion",
                        "Tuberose",
                        "Ranunculus"
                      ]
['Rose', 'Sunflower', 'Poppy', 'Pansy', 'Tulip', 'Fuchsia', 'Cyclamen', 'Lavender', 'Daisy', 'Jasmine', 'Hydrangea', 'Hyacinth', 'Peony', 'Dahlia', 'Dandelion', 'Tuberose', 'Ranunculus']

>>> nested_data_structures = [
                                 {"fish": "gold", "monkey": "brown", "parrot" : "grey"},
                                 ("fish", "mammal", "bird"),
                                 ['water', 'jungle', 'sky']
                             ]
[{"fish": "gold", "monkey": "brown", "parrot" : "grey"}, ("fish", "mammal", "bird"), ['water', 'jungle', 'sky']]
```

The `list()` constructor can be used empty or with an _iterable_ as an argument. Elements in the iterable are cycled through by the constructor and added to the list in order:

```python
>>> no_elements = list()
[]

> #the tuple is unpacked and each element is added

Reviewer: Comments should start with a # and a single space in all the code blocks.

>>> multiple_elements_tuple = list(("Parrot", "Bird", 334782))
["Parrot", "Bird", 334782]

#the set is unpacked and each element is added
>>> multiple_elements_set = list({2, 3, 5, 7, 11})
[2,3,5,7,11]
```

Results when using a list constructor with a string or a dict may be surprising:

````python

#string elements (Unicode code points) are iterated through and added *individually*
>>> multiple_elements_string = list("Timbuktu")
['T', 'i', 'm', 'b', 'u', 'k', 't', 'u']

>>> multiple_code_points_string = list('अभ्यास')
['अ', 'भ', '्', 'य', 'ा', 'स']

"""
The iteration default for dictionaries is over the keys.
"""
source_data = {"fish": "gold", "monkey": "brown"}
>>> multiple_elements_dict_1 = list(source_data)
['fish', 'monkey']

> 

Reviewer: Markdown code block need to be closed here

```python

>>> one_element = list(16)
Traceback (most recent call last):
   File "<stdin>", line 1, in <module>
   TypeError: 'int' object is not iterable

>>> one_element_from_iterable = list((16,))
[16]
````

## Accessing elements

Items inside lists (_like the sequence types `string` and `tuple`_), can be accessed via 0-based index and _bracket notation_. Indexes can be from **`left`** --> **`right`** (_starting at zero_) or **`right`** --> **`left`** (_starting at -1_).

| **0** | **1** | **2** | **3** | **4** | **5** |\
 -------------------------\
 | P | y | t | h | o | n |\
 -------------------------\
 |_**-6**_ |_**-5**_ |_**-4**_ |_**-3**_ |_**-2**_ |_**-1**_ | <----

```python

>>> breakfast_foods = ["Oatmeal", "Fruit Salad", "Eggs", "Toast"]

#Oatmeal is at index 0 or index -4
>>> first_breakfast_food = breakfast_foods[0]
'Oatmeal'

>>> first_breakfast_food = breakfast_foods[-4]
'Oatmeal'
```

The method `.pop()` can be used to both remove and return a value at a given index:

```python

>>> breakfast_foods = ["Oatmeal", "Fruit Salad", "Eggs", "Toast"]

> #Fruit Salad is at index 1 or index -3

Reviewer: Same for all comments in code blocks:
"Comments should start with a # and a single space." Pep8

                                 ("fish", "mammal", "bird"),
                                 ['water', 'jungle', 'sky']
                             ]
[{"fish": "gold", "monkey": "brown", "parrot" : "grey"}, ("fish", "mammal", "bird"), ['water', 'jungle', 'sky']]
```

The `list()` constructor can be used empty or with an _iterable_ as an argument. Elements in the iterable are cycled through by the constructor and added to the list in order:

```python
>>> no_elements = list()
[]

#the tuple is unpacked and each element is added
>>> multiple_elements_tuple = list(("Parrot", "Bird", 334782))
["Parrot", "Bird", 334782]

#the set is unpacked and each element is added
>>> multiple_elements_set = list({2, 3, 5, 7, 11})
[2,3,5,7,11]
```

Results when using a list constructor with a string or a dict may be surprising:

```python

#string elements (Unicode code points) are iterated through and added *individually*
>>> multiple_elements_string = list("Timbuktu")
['T', 'i', 'm', 'b', 'u', 'k', 't', 'u']

>>> multiple_code_points_string = list('अभ्यास')
['अ', 'भ', '्', 'य', 'ा', 'स']

"""
The iteration default for dictionaries is over the keys.

> To include both keys and values in a list made from a dictionary, use dict.items(),
> which will return a list of (key, value) tuples.
> """
> source_data = {"fish": "gold", "monkey": "brown"}
> >>> multiple_elements_dict_1 = list(source_data)
> ['fish', 'monkey']
> 
> >>> multiple_elements_dict_2 = list(source_data.items())
> [('fish', 'gold'), ('monkey', 'brown')]
> ```

Author: ```suggestion
"""
source_data = {"fish": "gold", "monkey": "brown"}
>>> multiple_elements_dict_1 = list(source_data)
['fish', 'monkey']
```
```

   TypeError: 'int' object is not iterable

>>> one_element_from_iterable = list((16,))
[16]
```

## Accessing elements

Items inside lists (_like the sequence types `string` and `tuple`_), can be accessed via 0-based index and _bracket notation_. Indexes can be from **`left`** --> **`right`** (_starting at zero_) or **`right`** --> **`left`** (_starting at -1_).

| **0** | **1** | **2** | **3** | **4** | **5** |\
 -------------------------\
 | P | y | t | h | o | n |\
 -------------------------\
 |_**-6**_ |_**-5**_ |_**-4**_ |_**-3**_ |_**-2**_ |_**-1**_ | <----

```python

>>> breakfast_foods = ["Oatmeal", "Fruit Salad", "Eggs", "Toast"]

#Oatmeal is at index 0 or index -4
>>> first_breakfast_food = breakfast_foods[0]
'Oatmeal'

>>> first_breakfast_food = breakfast_foods[-4]
'Oatmeal'
```

The method `.pop()` can be used to both remove and return a value at a given index:

```python

>>> breakfast_foods = ["Oatmeal", "Fruit Salad", "Eggs", "Toast"]

#Fruit Salad is at index 1 or index -3
>>> breakfast_foods = ["Oatmeal", "Fruit Salad", "Eggs", "Toast"]
>>> fruit_on_the_side = breakfast_foods.pop(-3)
'Fruit Salad'

>>> print(breakfast_foods)
['Oatmeal', 'Eggs', 'Toast']

```

> The method `.insert()` can be used to add an element at a specific position. The index given is the element _*before which to insert*_. `list.insert(0,element)` will insert at the front of the list and `list.insert(len(list), element)` is the same as calling `list.append(element)`.

Author: ```suggestion
The method `.insert()` can be used to add an element at a specific position. The index given is the element _*before which to insert*_. `list.insert(0,element)` will insert at the front of the list and `list.insert(len(list), element)` is the equivalent of calling `list.append(element)`.
```

>>> breakfast_foods.insert(3,"Bacon")
>>> print(breakfast_foods)
['Oatmeal', 'Fruit Salad', 'Eggs', 'Bacon', 'Toast']

#adding coffee in the first position
>>> breakfast_foods.insert(0, "Coffee")
>>> print(breakfast_foods)
['Coffee', 'Oatmeal', 'Fruit Salad', 'Eggs', 'Bacon', 'Toast']
```

## Working with lists

Lists supply an _iterator_, and can be looped through/over in the same manner as other _sequence types_:

```python

>>> colors = ["Orange", "Green", "Grey", "Blue"]
>>> for item in colors:
...     print(item)
...
Orange
Green
Grey
Blue

>>> numbers_to_cube = [5, 13, 12, 16]
>>> for number in numbers_to_cube:
...     print(number*3)
...
15
39
36
48

```

One common way to compose a list of values is to use `list.append()` with a loop:

```python

>>> cubes_to_1000 = []

>>> for number in range(11):
...    cubes_to_1000.append(number**3)

>>> print(cubles_to_1000)
[0, 1, 8, 27, 64, 125, 216, 343, 512, 729, 1000]
```

Lists can be combined via various techniques:

```python
# using the plus + operator unpacks each list and creates a new list, but it is not efficent.
>>> new_via_concatenate = ["George", 5] + ["cat", "Tabby"]
["George", 5, "cat", "Tabby"]

> #likewise, using the multiplication operator * is the equvilent of using + n times

Author: ```suggestion
#likewise, using the multiplication operator * is the equivalent of using + n times
```

>>> first_group = ["cat", "dog", "elephant"]

>>> multiplied_group = first_group * 3
['cat', 'dog', 'elephant', 'cat', 'dog', 'elephant', 'cat', 'dog', 'elephant']

# a more efficent method of combining 2 lists is to use slice asignment or appending in a loop
# by mutating one of the original lists
first_one = ["cat", "Tabby"]
second_one = ["George", 5]

# this assigns the second list to index 0 in the first list
>>> first_one[0:0] = second_one
>>> first_one
["George", 5, "cat", "Tabby"]

# this loops through the first list and appends it's items to the end of the second list
>>> for item in first_one:
>>>      second_one.append(item)
...
>>> print(second_one)
["George", 5, "cat", "Tabby"]
```

## Related data types

> Lists are often used as _stacks_ and _queues_ -- although their underlying implementation makes prepending and inserting slow. The [collections][collections] module offers a [deque][deque] variant optimized for fast appends and pops from either end that is implemented as a [doubly linked list][doubly linked list]. Nested lists are also used or to model small _matrices_ -- although the [Numpy][numpy] and [Pandas][pandas] libraries are much more robust for efficient matrix and tabular data manipulation. The collections module also provides a `UserList` type that can be customized to fit specialized list needs.

Author: ```suggestion
Lists are often used as _stacks_ and _queues_ -- although their underlying implementation makes prepending and inserting slow. The [collections][collections] module offers a [deque][deque] variant optimized for fast appends and pops from either end that is implemented as a [doubly linked list][doubly linked list]. Nested lists are also used to model small _matrices_ -- although the [Numpy][numpy] and [Pandas][pandas] libraries are much more robust for efficient matrix and tabular data manipulation. The collections module also provides a `UserList` type that can be customized to fit specialized list needs.
```

> In Python, a [list][list] is an mutable collection of items in _sequence_. Like most collections (_see the built-ins [`tuple`][tuple], [`dict`][dict] and [`set`][set]_), lists can hold reference to any (or multiple) data type(s) - including other lists. Like any [sequence][sequence type], items are referenced by 0-based index number, and can be copied in whole or in part via _slice notation_. Lists support all [common sequence operations][common sequence operations], as well as [mutable sequence operations][mutable sequence operations] like `.append()` and `.reverse()` They can be iterated over in a loop by using the `for item in` construct.

Reviewer: In Python, a [list][list] is a mutable collection of items in _sequence_. Like most collections (_see the built-ins [`tuple`][tuple], [`dict`][dict] and [`set`][set]_), lists can hold reference to any (or multiple) data type(s) - including other lists. Like any [sequence][sequence type], items are referenced by 0-based index number, and can be copied in whole or in part via _slice notation_. Lists support all [common sequence operations][common sequence operations], as well as [mutable sequence operations][mutable sequence operations] like `.append()` and `.reverse()`. They can be iterated over in a loop by using the `for item in` construct.

```python
>>> no_elements = list()
[]

#the tuple is unpacked and each element is added
>>> multiple_elements_tuple = list(("Parrot", "Bird", 334782))
["Parrot", "Bird", 334782]

#the set is unpacked and each element is added
>>> multiple_elements_set = list({2, 3, 5, 7, 11})
[2,3,5,7,11]
```

Results when using a list constructor with a string or a dict may be surprising:

```python

#string elements (Unicode code points) are iterated through and added *individually*
>>> multiple_elements_string = list("Timbuktu")
['T', 'i', 'm', 'b', 'u', 'k', 't', 'u']

>>> multiple_code_points_string = list('अभ्यास')
['अ', 'भ', '्', 'य', 'ा', 'स']

"""
The iteration default for dictionaries is over the keys.
To include both keys and values in a list made from a dictionary, use dict.items(),
which will return a list of (key, value) tuples.
"""
source_data = {"fish": "gold", "monkey": "brown"}
>>> multiple_elements_dict_1 = list(source_data)
['fish', 'monkey']

>>> multiple_elements_dict_2 = list(source_data.items())
[('fish', 'gold'), ('monkey', 'brown')]

> ```

Author: I go back and forth here.  Since I have examples of other data types, I though I would include dicts as well.  But dicts iterate differently -- hence the note.  I'm on the fence tho.  We can eliminate it.

> Because the constructor will only take _iterables_ (or nothing) as arguments, objecst that are _not_ iterable will throw a sytax error. Consequently, it is much easier to create a one-item list via the literal method.

Reviewer: ```suggestion
Because the constructor will only take _iterables_ (or nothing) as arguments, objects that are _not_ iterable will throw a syntax error. Consequently, it is much easier to create a one-item list via the literal method.
```

> ```python
> 
> >>> one_element = list(16)
> File "<stdin>", line 1
>     >>> one_element = list(16)
>     ^
> SyntaxError: invalid syntax

Author: ```suggestion
```python

>>> one_element = list(16)
Traceback (most recent call last):
   File "<stdin>", line 1, in <module>
   TypeError: 'int' object is not iterable
```

>>> one_element_from_iterable = list((16,))
[16]
```

## Accessing elements

Items inside lists (_like the sequence types `string` and `tuple`_), can be accessed via 0-based index and _bracket notation_. Indexes can be from **`left`** --> **`right`** (_starting at zero_) or **`right`** --> **`left`** (_starting at -1_).

| **0** | **1** | **2** | **3** | **4** | **5** |\
 -------------------------\
 | P | y | t | h | o | n |\
 -------------------------\
 |_**-6**_ |_**-5**_ |_**-4**_ |_**-3**_ |_**-2**_ |_**-1**_ | <----

```python

>>> breakfast_foods = ["Oatmeal", "Fruit Salad", "Eggs", "Toast"]

#Oatmeal is at index 0 or index -4
>>> first_breakfast_food = breakfast_foods[0]
'Oatmeal'

>>> first_breakfast_food = breakfast_foods[-4]
'Oatmeal'
```

The method `.pop()` can be used to both remove and return a value at a given index:

```python

>>> breakfast_foods = ["Oatmeal", "Fruit Salad", "Eggs", "Toast"]

#Fruit Salad is at index 1 or index -3
>>> breakfast_foods = ["Oatmeal", "Fruit Salad", "Eggs", "Toast"]
>>> fruit_on_the_side = breakfast_foods.pop(-3)
'Fruit Salad'

>>> print(breakfast_foods)
['Oatmeal', 'Eggs', 'Toast']

```

> The method `.insert()` can be used to add an element at a specific position. The index given is the element _*before which to insert*_. `list.insert(0,element)` will insert at the front of the list and `list.insert(len(list), element)` is the same as calling `list.append(element)`.

Reviewer 2: _Is_ `list.insert(len(list), element)` actually the same as `list.append(element)`? Or is just _functionally_ the same?

Author: Good question.  I copied it out of the docs, but to be honest, I don't actually _know_.  https://docs.python.org/3/tutorial/datastructures.html

```python

breakfast_foods = ["Oatmeal", "Fruit Salad", "Eggs", "Toast"]

#adding bacon to the mix before index 3 or index -1
>>> breakfast_foods.insert(3,"Bacon")
>>> print(breakfast_foods)
['Oatmeal', 'Fruit Salad', 'Eggs', 'Bacon', 'Toast']

#adding coffee in the first position
>>> breakfast_foods.insert(0, "Coffee")

> print(breakfast_foods)

Reviewer 2: ```suggestion
>>> print(breakfast_foods)
```

['Coffee', 'Oatmeal', 'Fruit Salad', 'Eggs', 'Bacon', 'Toast']
```

## Working with lists

Lists supply an _iterator_, and can be looped through/over in the same manner as other _sequence types_:

```python

>>> colors = ["Orange", "Green", "Grey", "Blue"]
>>> for item in colors:
...     print(item)
...
Orange
Green
Grey
Blue

>>> numbers_to_cube = [5, 13, 12, 16]

> for number in numbers_to_cube:
>     print(number*3)

Reviewer 2: ```suggestion
>>> for number in numbers_to_cube:
...     print(number*3)
```

...
15
39
36
48

```

One common way to compose a list of values is to use `list.append()` with a loop:

```python

> >>> cubes_to 1000 = []

Reviewer 2: ```suggestion
>>> cubes_to_1000 = []
```

...    cubes_to_1000.append(number**3)

>>> print(cubles_to_1000)
[0, 1, 8, 27, 64, 125, 216, 343, 512, 729, 1000]
```

Lists can be combined via various techniques:

```python
# using the plus + operator unpacks each list and creates a new list, but it is not efficent.
>>> new_via_concatenate = ["George", 5] + ["cat", "Tabby"]
["George", 5, "cat", "Tabby"]

#likewise, using the multiplication operator * is the equvilent of using + n times
>>> first_group = ["cat", "dog", "elephant"]

>>> multiplied_group = first_group * 3
['cat', 'dog', 'elephant', 'cat', 'dog', 'elephant', 'cat', 'dog', 'elephant']

# a more efficent method of combining 2 lists is to use slice asignment or appending in a loop
# by mutating one of the original lists
first_one = ["cat", "Tabby"]
second_one = ["George", 5]

# this assigns the second list to index 0 in the first list
>>> first_one[0:0] = second_one
>>> first_one
["George", 5, "cat", "Tabby"]

# this loops through the first list and appends it's items to the end of the second list
>>> for item in first_one:
>>>      second_one.append(item)
...
>>> print(second_one)
["George", 5, "cat", "Tabby"]
```

## Related data types

> Lists are often used as _stacks_ and _queues_ -- although their underlying implementation makes prepending and inserting slow. The [collections][collections] module offers a [deque][deque] variant optimized for fast appends and pops from either end that is implemented as a [doubly linked list][doubly linked list]. Nested lists ara also used or to model small _matrcies_ -- although the [Numpy][numpy] and [Pandas][pandas] libraries are much more robust for efficent matrix and tabular data manipulation. The collections module also provides a `UserList` type that can be customized to fit specilized list needs.

Reviewer 2: ```suggestion
Lists are often used as _stacks_ and _queues_ -- although their underlying implementation makes prepending and inserting slow. The [collections][collections] module offers a [deque][deque] variant optimized for fast appends and pops from either end that is implemented as a [doubly linked list][doubly linked list]. Nested lists are also used or to model small _matrices_ -- although the [Numpy][numpy] and [Pandas][pandas] libraries are much more robust for efficient matrix and tabular data manipulation. The collections module also provides a `UserList` type that can be customized to fit specialized list needs.
```
